v-next/hardhat/src/internal/cli/main.ts

```diff
@@ -265,3 +269,192 @@ function getTaskFromCliArguments(
 
   return task;
 }
+
+function parseTaskArguments(
+  cliArguments: string[],
+  usedCliArguments: boolean[],
+  task: Task,
+): Record<string, any> {
+  const taskArguments: Record<string, unknown> = {};
+
+  parseNamedParameters(cliArguments, usedCliArguments, task, taskArguments);
+
+  parsePositionalAndVariadicParameters(
+    cliArguments,
+    usedCliArguments,
+    task,
+    taskArguments,
+  );
+
+  usedCliArguments.forEach((arg, index) => {
+    if (!arg) {
+      throw new HardhatError(HardhatError.ERRORS.ARGUMENTS.UNUSED_ARGUMENT, {
+        value: cliArguments[index],
+      });
+    }
+  });
+
+  return taskArguments;
+}
+
+function parseNamedParameters(
+  cliArguments: string[],
+  usedCliArguments: boolean[],
+  task: Task,
+  taskArguments: Record<string, any>,
+) {
+  for (let i = 0; i < cliArguments.length; i++) {
+    if (usedCliArguments[i]) {
+      continue;
+    }
+
+    if (cliArguments[i] === "--") {
+      // A standalone '--' is ok because it is used to separate CLI tool arguments from script arguments,
+      // ensuring the tool passes subsequent options directly to the script.
+      continue;
+    }
+
+    const arg = cliArguments[i];
+
+    if (arg.startsWith("--") === false) {
+      continue;
+    }
+
+    const paramName = kebabToCamelCase(arg.substring(2));
+    const paramInfo = task.namedParameters.get(paramName);
+
+    if (paramInfo === undefined) {
+      throw new HardhatError(
+        HardhatError.ERRORS.ARGUMENTS.UNRECOGNIZED_NAMED_PARAM,
+        {
+          parameter: arg,
+        },
+      );
+    }
+
+    usedCliArguments[i] = true;
+
+    if (paramInfo.parameterType === ParameterType.BOOLEAN) {
+      if (
+        usedCliArguments[i + 1] === false &&
+        ["true", "false"].includes(cliArguments[i + 1])
+      ) {
+        // The flag could be follow by the boolean value
+        taskArguments[paramName] = formatParameterValue(
+          cliArguments[i + 1],
+          ParameterType.BOOLEAN,
+          paramName,
+        );
+
+        usedCliArguments[i + 1] = true;
+        continue;
+      }
+
+      // If the flag is not followed by a boolean value, then the flag itself means that the value is true
+      taskArguments[paramName] = true;
+      continue;
+    }
+
+    // The value immediately following a named parameter (if the parameter does not behave as a flag)
+    // is the parameter's value; otherwise, it's an error
+    if (usedCliArguments[i + 1] === false) {
```

Comment:
Also here, I think this can be out of bounds

Reply:
This actually works correctly, but it looks like unintended behavior.

Reply:
Before I had a out of bound check, but then I realized that, as you mention, this work correctly. So I decided to remove code that was "useless". Do you think that being more explicit would be better?

Reply:
[Done in this commit](https://github.com/NomicFoundation/hardhat/pull/5300/commits/e3e958a3ed540e89742f3fa6af3a3e6e62f3e351)